core/services/functions/connector_handler.go

```diff
@@ -92,24 +121,35 @@ func (h *functionsConnectorHandler) HandleGatewayMessage(ctx context.Context, ga
 			return
 		}
 		h.handleSecretsSet(ctx, gatewayId, body, fromAddr)
+	case functions.MethodHeartbeat:
+		h.handleHeartbeat(ctx, gatewayId, body, fromAddr)
 	default:
 		h.lggr.Errorw("unsupported method", "id", gatewayId, "method", body.Method)
 	}
 }
 
 func (h *functionsConnectorHandler) Start(ctx context.Context) error {
 	return h.StartOnce("FunctionsConnectorHandler", func() error {
+		h.serviceContext, h.serviceCancel = context.WithCancel(context.Background())
 		if err := h.allowlist.Start(ctx); err != nil {
 			return err
 		}
-		return h.subscriptions.Start(ctx)
+		if err := h.subscriptions.Start(ctx); err != nil {
+			return err
+		}
+		h.shutdownWaitGroup.Add(1)
+		go h.reportLoop()
+		return nil
 	})
 }
 
 func (h *functionsConnectorHandler) Close() error {
 	return h.StopOnce("FunctionsConnectorHandler", func() (err error) {
+		h.serviceCancel()
+		close(h.chStop)
```

Comment:
I think you can get rid of `chStop` and use `h.serviceContext`.

Reply:
I remember that we tried to do it in Listener a while ago and @jmank88 recommended to keep it. I don't remember why. @jmank88 can we just use serviceContext.Done() instead of a separate stop channel?

Reply:
`context.Context` isn't meant to be stored on structs (we aren't enforcing this yet, but we might via a golangci-lint rule), but the `StopChan` type was added to make it easy to spawn them. For now, either way is fine - whatever is simplest to align things.

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

https://pkg.go.dev/context

Reply:
Switching to StopChan, thanks.